### ISSUE TITLE
Add pipelining in Samtools-Sort

### ISSUE THREAD

**Issue by quito418** [samtools/samtools#1732]
#### Is your feature request related to a problem? Please specify.
Samtools Sort is often used with Linux pipeline, piped to alignment software (e.g. BWA-MEM).

When Samtools-Sort is piped with alignment software, the memory buffer is first filled in by alignment software. After the memory buffer is full, it writes the memory buffer into a temporary file. While the memory buffer is being dumped, Samtools-Sort does not read input from alignment software, which makes the alignment software `write function` to hang (this becomes significant when using a large memory buffer in Samtools-Sort).

#### Describe the solution you would like.

If we apply pipelining to Samtools-sort, we can resolve this issue. (When one pipeline is writing, make another pipeline to read and sort)

I tested whether the pipeline version works (https://github.com/quito418/samtools)  and confirmed it resolves the `write hang` problem.

Test setting: (16 physical core, 32thread machine, piped with BWA-MEM2, HDD disk, command: `bwa-mem2 mem -t 32  ref fasta1 fasta2 | samtools sort -m 1G -@ 10 -`)
Result: Disk I/O time in BWA-MEM2 reduces from 1350 to 300

(left is using pipeline sort, right is current develop branch samtools)
![image](https://user-images.githubusercontent.com/26645270/195083959-60e314c9-76e3-40e6-a13d-dc8fe71e4a43.png)

**Comment by jkbonfield** [comment]
To test if this is a bottleneck, you may wish to add various `mbuffer` parameter sets inbetween bwa and samtools to see if adding an additional memory buffer improves the throughput.

**Comment by daviesrob** [comment]
Overlapping reading the input and sorting is the next improvement that I'd like to make to samtools sort.  The main issue is that reserving memory to read into means it's likely to write more temporary files than it otherwise would have, which could slow the merging part of the process. Although if the memory is only divided into two, the situation will mostly still be better than it was before  #1701 was merged, when each sort thread wrote to a separate temp file.

I've had a look at your branch.  It's a little messy, but I think I've got a rough idea of how it works.  I think I'd do it slightly differently, with a thread that only does the reading part and one that only does writing, with buffers passed to-and-fro between them.  I think that be slightly simpler than the branch, which starts two threads that both do a read,write loop; and uses mutexes and conditionals to keep the reading and writing bits out of phase.  I'd have to try it though to be sure.
